### PR TITLE
More effective solve_poly_system()

### LINKED ISSUE #14805: More effective backsubstitution in solve_poly_system()
The solve_poly_system function, which is used for solving systems of polynomial equations, currently implements the backsubstitution method: construct groebner basis w.r.t. the lexicographic ordering and then use backsubstitution to find all solutions, starting with the univariate basis polynomial. If you look into the implementation of solve_poly_system, you can see that at each step of the backsubstitution, the groebner basis is computed again from scratch. Computation of groebner basis can be very expensive and you don't want to do it more than once if it's not necessary. In this case, it is quite easy to implement the backsubstitution method in a way that constructs the groebner basis only once.

So unless there is a reason why the groebner basis is computed from scratch at each step, I suggest I could fix this.

### PR DESCRIPTION
**Pull request by filip-misun**
The function solve_poly_system() finds solutions to a system of polynomial
equations. This is implemented using the elimination and backsubstitution:
we construct the groebner basis w.r.t. the lexicographic ordering and then we
recursively eliminate variables, starting with the univariate polynomial in
the basis. In the previous implementation, the groebner basis was computed
from scratch at each step of elimination. This is not necessary and it makes
the the solve_poly_system() very ineffective, since computation of groebner
basis can be very expensive(). In this commit, this function is rewritten so
that the groebner basis is computed only once at the beginning. I also made
some refactoring and added some comments too, making the implementation
more readable (hopefully). Fixes #14805.

### CODE REVIEW

**Comment by Abdullahjavednesar** [comment]
Can you add some test cases.

**Comment by filip-misun** [comment]
Well, I didn't add any new functionality. I didn't even fix a bug. I just made the solve_poly_system() function less time consuming, did some refactoring and added comments. Do I need to add a new test case anyway?

**Comment by smichr** [comment]
The changes you have made are breaking existing tests which means (I think) that the recomputation of basis needs to be done adaptively (read, "more than once"). What are your conclusions about the failures?

**Comment by skirpichev** [comment]
Where this algorithm described?

**Comment by filip-misun** [comment]
The algorithm is the same as before, just the implementation is different. A brief description of the algorithm is in the comments and an appropriate literature is cited.

I am still convinced that recomputation of basis is not necessary. There are other problems with the implementation. Whenever an univariate polynomial in the basis is solved, the solutions are substituted into the remaining polynomials. It might happen, that some polynomials in the basis become constant because of this substitution (i.e. there are no free variables, for which we solve the system, in this polynomial). I have to test if these constant polynomials are zero. Consider the following example:
```
x,y = symbols('x,y')
f1 = x**2 - 1
f2 = y**2 - 1
f3 = x*y + x + y + 1
solve_poly_system([f1,f2,f3], x,y)
```
The system [f1,f2,f3] is a groebner basis. The algorithm is going to make the following backsubstitution. First, it solves f1, finds solution x=1 and substitutes x=1 in f2 and f3. Then it solves f2, finds solution y=1. Although (x=1,y=1) satisfies polynomials f1,f2, we are not done. We have to check, if (x=1,y=1) satisfies f3, i.e. we have to check if f3=0 after substitution. We can see that in this case, f3 evaluates to 4, so (x=1,y=1) is not a solution.

So I have an expression `p=f3.subs([(x,1),(y,1)])`. How do I check if p is zero? Currently, I am using `p.simplify() == 0`. However, this doesn't always work and this is the reason why my implementation fails in sympy/solvers/tests/test_polysys.py", line 36. In this instance, simplify() has trouble to recognize the fact, that
```
-3*sqrt(12*sqrt(2) + 17) - 2*sqrt(2)*sqrt(12*sqrt(2) + 17) + 12*sqrt(2) + 17 == 0
```
For this reason, one solution is lost.

To summarize, I have to check if a given expression is equal to zero. What can I do, when even simplify() fails? Can anybody help me with this.

By the way, it seems to me (but I might be wrong) that the current implementation of solve_poly_system() suffers from the same problem. In this particular test case, I think that it might pass by just "luckily" eliminating the variables in different order and therefore avoiding the complex expression above.

**Comment by filip-misun** [comment]
When I have more time, I will analyse the second failed test as well. But I can tell right off the bat that the test doesn't look quite "fair". Look at it:
```
def test_issue_12114():
    a, b, c, d, e, f, g = symbols('a,b,c,d,e,f,g')
    terms = [1 + a*b + d*e, 1 + a*c + d*f, 1 + b*c + e*f,
             g - a**2 - d**2, g - b**2 - e**2, g - c**2 - f**2]
    s = solve(terms, [a, b, c, d, e, f, g], dict=True)
    assert s == [{a: -sqrt(-f**2 - 1), b: -sqrt(-f**2 - 1),
                  c: -sqrt(-f**2 - 1), d: f, e: f, g: -1},
                 {a: sqrt(-f**2 - 1), b: sqrt(-f**2 - 1),
                  c: sqrt(-f**2 - 1), d: f, e: f, g: -1},
                 {a: -sqrt(3)*f/2 - sqrt(-f**2 + 2)/2,
                  b: sqrt(3)*f/2 - sqrt(-f**2 + 2)/2, c: sqrt(-f**2 + 2),
                  d: -f/2 + sqrt(-3*f**2 + 6)/2,
                  e: -f/2 - sqrt(3)*sqrt(-f**2 + 2)/2, g: 2},
                 {a: -sqrt(3)*f/2 + sqrt(-f**2 + 2)/2,
                  b: sqrt(3)*f/2 + sqrt(-f**2 + 2)/2, c: -sqrt(-f**2 + 2),
                  d: -f/2 - sqrt(-3*f**2 + 6)/2,
                  e: -f/2 + sqrt(3)*sqrt(-f**2 + 2)/2, g: 2},
                 {a: sqrt(3)*f/2 - sqrt(-f**2 + 2)/2,
                  b: -sqrt(3)*f/2 - sqrt(-f**2 + 2)/2, c: sqrt(-f**2 + 2),
                  d: -f/2 - sqrt(-3*f**2 + 6)/2,
                  e: -f/2 + sqrt(3)*sqrt(-f**2 + 2)/2, g: 2},
                 {a: sqrt(3)*f/2 + sqrt(-f**2 + 2)/2,
                  b: -sqrt(3)*f/2 + sqrt(-f**2 + 2)/2, c: -sqrt(-f**2 + 2),
                  d: -f/2 + sqrt(-3*f**2 + 6)/2,
                  e: -f/2 - sqrt(3)*sqrt(-f**2 + 2)/2, g: 2}]
```
The output of the solve() function is compared with the expected solution using "==". Whenever comparing expressions p1 and p2 for equality, shouldn't we rather use `(p1-p2).simplify() == 0`, or something like that? If `p1 == p2` returns `False`, it doesn't mean expressions p1 and p2 are not equal. Moreover, notice that the polynomial system on the input is positive-dimensional. So there are infinitely many solutions and therefore, the solutions are parametrized through a parameter. However, why does the test insist that the solutions have to be parametrized through symbol f? It shouldn't matter what parameter I choose.

**Comment by skirpichev** [comment]
> an appropriate literature is cited.

I see, Cox97 reference is appropriate here.

> To summarize, I have to check if a given expression is equal to zero. What can I do, when even simplify() fails?

In this particular case you might exploit ``minimal_polynomial(expr, x)``.  Beware, that for given expression it raises NotImplementedError (default algorithm has some bug), only groebner-based algorithm (``compose=False`` in sympy or ``method='groebner'`` in diofant) gives ``x``.

> However, why does the test insist that the solutions have to be parametrized through symbol f? It shouldn't matter what parameter I choose.

Because sympy algorithms (well, some bunch of heuristics) choose parameters this way.  But if you have an idea how to rewrite this test to be independent from the parametrization - feel free to do.  Be sure to test your test, i.e. it should fail before fix for #12114 was introduced.

**Comment by smichr** [comment]
> simplify() has trouble to recognize the fact, that
> 
> -3*sqrt(12*sqrt(2) + 17) - 2*sqrt(2)*sqrt(12*sqrt(2) + 17) + 12*sqrt(2) + 17 == 0

The expensive (and not guaranteed to succeed) `equals` function can help:

```
>>> (-3*sqrt(12*sqrt(2) + 17) - 2*sqrt(2)*sqrt(12*sqrt(2) + 17) + 12*sqrt(2) + 17).equals(0)
True
```

**Comment by smichr** [comment]
> How do I check if p is zero?

Since groebner is already expensive, using another expensive function that will give more success might be a solution (but a general method of [proving A = B is not possible](https://www.math.upenn.edu/~wilf/AeqB.pdf)).

Use `p.equals(0)`.
